File: python/service.py
from fastapi import HTTPException
from python.models import Category, Expense
from python.schemas import CreateExpensePayload, GetExpensesPayload
from sqlalchemy.orm import Session
from sqlalchemy import func
from datetime import datetime

def upsert_expense(expense: CreateExpensePayload, session: Session):
    expense.date_of_expense = datetime.strptime(expense.date_of_expense, "%Y-%m-%d").date()
    existing_expense: Expense = session.query(Expense).filter(Expense.id == expense.expense_id).first()
    category: Category = session.query(Category).filter(func.lower(Category.name) == expense.category_name.lower()).first()
    if not category:
        raise HTTPException(status_code=404, detail="Category not found, please try another Category")
    if existing_expense:
        # Update existing expense
        existing_expense.name = expense.name
        existing_expense.description = expense.description
        existing_expense.amount_in_cents = (expense.amount) * 100
        existing_expense.date_of_expense = expense.date_of_expense
        existing_expense.category_id = category.id
        try:
            session.add(existing_expense)
            session.commit()
            session.refresh(existing_expense)
            return existing_expense
        except Exception as e:
            session.rollback()
            raise HTTPException(status_code=500, detail="Error updating expense: " + str(e))
    else:
        try:
            expense = Expense(
                name=expense.name,
                description=expense.description,
                amount_in_cents=(expense.amount) * 100,
                date_of_expense=expense.date_of_expense,
                category_id=category.id,
                user_id=expense.user_id,
                is_archived=False
            )
            session.add(expense)
            session.commit()
            session.refresh(expense)
            return expense
        except Exception as e:
            session.rollback()
            raise HTTPException(status_code=500, detail="Error creating expense: " + str(e))

def delete_expense(expense_id: int, session: Session):
    expense = session.query(Expense).filter(Expense.id == expense_id).first()
    if not expense:
        raise HTTPException(status_code=404, detail="Expense not found")
    
    session.delete(expense)
    session.commit()
    return {"detail": "Expense deleted successfully"}

def get_expenses(payload: GetExpensesPayload, session: Session):
    filters = []
    if payload.start_date:
        payload.start_date = datetime.strptime(payload.start_date, "%Y-%m-%d").replace(hour=0, minute=0, second=0, microsecond=0)
        filters.append(Expense.date_of_expense >= payload.start_date)
    if payload.end_date:
        payload.end_date = datetime.strptime(payload.end_date, "%Y-%m-%d").replace(hour=23, minute=59, second=59, microsecond=999999)
        filters.append(Expense.date_of_expense <= payload.end_date)
    if payload.category_name:
        category = session.query(Category).filter(Category.name == payload.category_name).first()
        if not category:
            raise HTTPException(status_code=404, detail="Category not found")
        filters.append(Expense.category_id == category.id)
    expenses = session.query(Expense, Category.name).join(Category, Category.id == Expense.category_id).filter(Expense.user_id == payload.user_id).filter(*filters).all()
    expenses_return = []
    for expense_data in expenses:
        expense: Expense = expense_data["Expense"]
        expense = {
            "id": expense.id,
            "name": expense.name,
            "description": expense.description,
            "amount_in_cents": expense.amount_in_cents,
            "date_of_expense": expense.date_of_expense,
            "category_id": expense.category_id,
            "user_id": expense.user_id,
            "is_archived": expense.is_archived
        }
        expense["category_name"] = expense_data[1]
        expense["amount"] = expense["amount_in_cents"] / 100
        expense["date_of_expense"] = expense["date_of_expense"].strftime("%Y-%m-%d")
        expenses_return.append(expense)
    return expenses_return

    
    

File: cpp/service.cpp
#include <SQLiteCpp/SQLiteCpp.h>
#include <crow/json.h>
#include <sstream>
#include <stdexcept>
#include <iostream>
#include "models.hpp"
#include "schemas.hpp"


Expense upsertExpense(SQLite::Database& db, const CreateExpensePayload& payload) {
    // Fetch Category
    SQLite::Statement category_query(db, "SELECT id FROM categories WHERE lower(name) = lower(?)");
    category_query.bind(1, payload.categoryName);

    if (!category_query.executeStep()) {
        throw std::runtime_error("Category not found");
    }
    int category_id = category_query.getColumn(0).getInt();
    Expense expense;
    int amount_in_cents = static_cast<int>(payload.amount * 100);
    if (payload.expenseId.has_value()) {
        // Update existing expense
        SQLite::Statement update_query(db,
            "UPDATE expenses SET name=?, description=?, amount_in_cents=?, date_of_expense=?, category_id=? WHERE id=?");
        
        update_query.bind(1, payload.name);
        update_query.bind(2, payload.description.value_or(""));
        update_query.bind(3, amount_in_cents);
        update_query.bind(4, payload.dateOfExpense);
        update_query.bind(5, category_id);
        update_query.bind(6, payload.expenseId.value());
        update_query.exec();

        expense.id = payload.expenseId.value();
    } else {
        SQLite::Statement insert_query(db,
            "INSERT INTO expenses (name, description, amount_in_cents, date_of_expense, category_id, user_id, is_archived) "
            "VALUES (?, ?, ?, ?, ?, ?, 0)");

        insert_query.bind(1, payload.name);
        insert_query.bind(2, payload.description.value_or(""));
        insert_query.bind(3, amount_in_cents);
        insert_query.bind(4, payload.dateOfExpense);
        insert_query.bind(5, category_id);
        insert_query.bind(6, payload.userId);
        insert_query.exec();

        expense.id = static_cast<int>(db.getLastInsertRowid());
    }

    expense.name = payload.name;
    expense.description = payload.description.value_or("");
    expense.amountInCents = amount_in_cents;
    expense.dateOfExpense = payload.dateOfExpense;
    expense.categoryId = category_id;
    expense.userId = payload.userId;
    expense.isArchived = false;

    return expense;
}

bool deleteExpense(SQLite::Database& db, int expense_id) {
    SQLite::Statement stmt(db, "DELETE FROM expenses WHERE id = ?");
    stmt.bind(1, expense_id);
    return stmt.exec() > 0;
}

std::vector<crow::json::wvalue> getExpenses(SQLite::Database& db, const GetExpensesPayload& payload) {
    std::ostringstream sql;
    sql << "SELECT e.id, e.name, e.description, e.amount_in_cents, e.date_of_expense, "
        << "e.category_id, e.user_id, e.is_archived, c.name "
        << "FROM expenses e "
        << "JOIN categories c ON e.category_id = c.id "
        << "WHERE e.user_id = ?";

    if (payload.startDate.has_value() && payload.startDate.value() != "") {
        sql << " AND e.date_of_expense >= '" << payload.startDate.value() << "'";
    }
    if (payload.endDate.has_value() && payload.endDate.value() != "") {
        sql << " AND e.date_of_expense <= '" << payload.endDate.value() << "23:59:59'";
    }
    if (payload.categoryName.has_value() && payload.categoryName.value() != "") {
        sql << " AND c.name = '" << payload.categoryName.value() << "'";
    }
    SQLite::Statement query(db, sql.str());
    query.bind(1, payload.userId);

    std::vector<crow::json::wvalue> results;

    while (query.executeStep()) {
        crow::json::wvalue item;
        item["id"] = query.getColumn(0).getInt();
        item["name"] = query.getColumn(1).getString();
        item["description"] = query.getColumn(2).getString();
        int amount_cents = query.getColumn(3).getInt();
        item["amount_in_cents"] = amount_cents;
        item["amount"] = std::round((amount_cents / 100.0) * 100) / 100.0;
        item["date_of_expense"] = query.getColumn(4).getString().substr(0, 10);
        item["category_id"] = query.getColumn(5).getInt();
        item["user_id"] = query.getColumn(6).getInt();
        item["is_archived"] = query.getColumn(7).getInt() != 0;
        item["category_name"] = query.getColumn(8).getString();
        results.push_back(item);
    }

    return results;
}
